File: parse_questions.py
import json
import re
import sys

def parse_file(filename):
    with open(filename, 'r', encoding='utf-8') as f:
        content = f.read()

    content = content.replace('\r\n', '\n')
    lines = content.split('\n')
    
    questions = []
    
    q_text = []
    options = {}
    
    for line in lines:
        s = line.strip()
        if not s:
            continue
            
        m = re.match(r'^([A-E])\s*[.)]\s*(.*)', s)
        if m:
            letter = m.group(1).upper()
            text = m.group(2)
            options[letter] = text
        elif re.match(r'^[A-E]{1,5}$', s) and len(options) > 0:
            questions.append({
                "question": "\n".join(q_text).strip(),
                "options": options.copy(),
                "answer": s
            })
            q_text = []
            options = {}
        else:
            if len(options) > 0:
                q_text = [s]
                options = {}
            else:
                q_text.append(s)
                
    return questions

if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Usage: python parse_questions.py <input.md> <output.json>")
        sys.exit(1)
        
    q = parse_file(sys.argv[1])
    
    # Filter out empty or invalid questions
    valid_q = []
    for x in q:
        if x["question"] and len(x["options"]) >= 2:
            valid_q.append(x)
            
    with open(sys.argv[2], 'w', encoding='utf-8') as f:
        json.dump(valid_q, f, ensure_ascii=False, indent=2)
    print(f"Parsed {len(valid_q)} valid questions out of {len(q)}")

File: parse_questions.js
const fs = require('fs');

function parseFile(filename) {
    const content = fs.readFileSync(filename, 'utf-8');
    const lines = content.replace(/\r\n/g, '\n').split('\n');
    
    const questions = [];
    let q_text = [];
    let options = {};
    
    for (const line of lines) {
        const s = line.trim();
        if (!s) continue;
        
        const optionMatch = s.match(/^([A-E])\s*[.)]\s*(.*)/i);
        const answerMatch = s.match(/^(?:Đáp án:\s*)?([A-E]{1,5})(?:\s*\.?)?(?:\s*\((.*)\))?$/i);
        
        if (optionMatch) {
            const letter = optionMatch[1].toUpperCase();
            const text = optionMatch[2];
            options[letter] = text;
        } else if (answerMatch && Object.keys(options).length > 0) {
            questions.push({
                question: q_text.join('\n').trim(),
                options: { ...options },
                answer: answerMatch[1].toUpperCase(),
                explanation: answerMatch[2] ? answerMatch[2].trim() : ""
            });
            q_text = [];
            options = {};
        } else {
            if (Object.keys(options).length > 0) {
                // If it's something like an alternative question that lacks an answer,
                // we'll just ignore it and start a new question block
                q_text = [s];
                options = {};
            } else {
                q_text.push(s);
            }
        }
    }
    
    return questions;
}

const input = process.argv[2];
const output = process.argv[3];

if (!input || !output) {
    console.error("Usage: node parse_questions.js <input.md> <output.json>");
    process.exit(1);
}

const q = parseFile(input);
const valid_q = q.filter(x => x.question && Object.keys(x.options).length >= 2);

const jsContent = "const allQuestions = " + JSON.stringify(valid_q, null, 2) + ";";
fs.writeFileSync(output, jsContent, 'utf-8');
console.log(`Parsed ${valid_q.length} valid questions out of ${q.length}`);
